components/region_cache_memory_engine/src/engine.rs

```diff
@@ -23,19 +35,87 @@ pub struct RegionMemoryEngine {
     data: [Arc<Skiplist<ByteWiseComparator>>; 3],
 }
 
+impl RegionMemoryEngine {
+    pub fn with_capacity(arena_size: usize) -> Self {
+        RegionMemoryEngine {
+            data: [
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
+impl Default for RegionMemoryEngine {
+    fn default() -> Self {
+        RegionMemoryEngine {
+            data: [
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
 impl Debug for RegionMemoryEngine {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        unimplemented!()
+        write!(f, "Region Memory Engine")
     }
 }
 
-type SnapshotList = BTreeMap<u64, u64>;
+// read_ts -> ref_count
+#[derive(Default)]
+struct SnapshotList(BTreeMap<u64, u64>);
+
+impl SnapshotList {
+    fn new_snapshot(&mut self, read_ts: u64) {
+        // snapshot with this ts may be granted before
+        let count = self.0.get(&read_ts).unwrap_or(&0) + 1;
+        self.0.insert(read_ts, count);
+    }
+
+    fn remove_snapshot(&mut self, read_ts: u64) {
+        let count = self.0.get_mut(&read_ts).unwrap();
+        assert!(*count >= 1);
```

Comment:
how about add more info, such as this read_ts

Reply:
What you mean? I dont get it.

Reply:
like this  `assert!(!resp.has_error(), "{:?}", read_ts);`

Reply:
No info is exposed in new_snaphsot, so adding this info here does not help.